pkg/controller/controller.go

```diff
@@ -286,22 +290,36 @@ func (p *csiProvisioner) checkDriverCapabilities(rc *requiredCapabilities) error
 	return nil
 }
 
-func makeVolumeName(prefix, pvcUID string, volumeNameUUIDLength int) (string, error) {
+func (p *csiProvisioner) makeVolumeName(pvcNamespace, pvcName, pvcUID string) (string, error) {
 	// create persistent name based on a volumeNamePrefix and volumeNameUUIDLength
-	// of PVC's UID
-	if len(prefix) == 0 {
+	// of PVC's UID; if enabled, utilize PVC namespace and name as well
+	if len(p.volumeNamePrefix) == 0 {
 		return "", fmt.Errorf("Volume name prefix cannot be of length 0")
 	}
 	if len(pvcUID) == 0 {
 		return "", fmt.Errorf("corrupted PVC object, it is missing UID")
 	}
-	if volumeNameUUIDLength == -1 {
-		// Default behavior is to not truncate or remove dashes
-		return fmt.Sprintf("%s-%s", prefix, pvcUID), nil
-	}
-	// Else we remove all dashes from UUID and truncate to volumeNameUUIDLength
-	return fmt.Sprintf("%s-%s", prefix, strings.Replace(string(pvcUID), "-", "", -1)[0:volumeNameUUIDLength]), nil
+	pvcUID = strings.Replace(pvcUID, "-", "", -1)
 
+	var fullName string
+	if p.volumeNamesReadable {
+		if len(pvcNamespace) == 0 {
+			return "", fmt.Errorf("corrupted PVC object, it is missing namespace")
+		}
+		if len(pvcName) == 0 {
+			return "", fmt.Errorf("corrupted PVC object, it is missing name")
+		}
+		pvcHash := md5.Sum([]byte(pvcNamespace + "/" + pvcName))
+		pvcNamespace = strings.Replace(pvcNamespace, "-", "", -1)
+		pvcName = strings.Replace(pvcName, "-", "", -1)
+		fullName = fmt.Sprintf("%s-%s-%x", pvcNamespace, pvcName, pvcHash)
```

Comment:
Both namespace name and PV names can be up 63 characters long. PV name is also limited to 63 characters. With a long namespace names, there may be no space for PVC name nor the hash and the resulting name is not unique.

Reply:
I didn't realize PV names were limited to 63 characters; however, even with that in mind, it's going to be desirable for some administrators (me, for example) to trade having to be careful about namespace/PVC naming for being able to find storage more easily.

I can certainly add more clarity to the help text on the flag.

Reply:
> Both namespace name and PV names can be up 63 characters long. PV name is also limited to 63 characters

I checked today and I was wrong, namespace name is has limit 63 characters, but PVC and PV names have 253 characters.

Still, with PVC a long PVC name, the code above produces PV name that will be rejected by API server. And it does not allow any space to add namespace name to a very long PVC name.

Reply:
> Both namespace name and PV names can be up 63 characters long. PV name is also limited to 63 characters

I checked today and I was wrong, namespace name is has limit 63 characters, but PVC name can have 253 characters. CSI specifies that any string, incl. volume name, can have up to 128 characters. So you don't have even place for full PVC name, not mentioning the namespace.